src/cart/src/cart/crt_internal_types.h

```diff
@@ -0,0 +1,276 @@
+ *
+ * 3. In addition, redistributions of modified forms of the source or binary
+ *    code must carry prominent notices stating that the original code was
+ *    changed and the date of the change.
+ *
+ *  4. All publications or advertising materials mentioning features or use of
+ *     this software are asked, but not required, to acknowledge that it was
+ *     developed by Intel Corporation and credit the contributors.
+ *
+ * 5. Neither the name of Intel Corporation, nor the name of any Contributor
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * This file is part of CaRT. It gives out the data types internally used by
+ * CaRT and not in other specific header files.
+ */
+
+#ifndef __CRT_INTERNAL_TYPES_H__
+#define __CRT_INTERNAL_TYPES_H__
+
+#define CRT_CONTEXT_NULL         (NULL)
+
+#include <arpa/inet.h>
+#include <ifaddrs.h>
+
+#include <gurt/list.h>
+#include <gurt/hash.h>
+#include <gurt/heap.h>
+#include <gurt/atomic.h>
+
+struct crt_hg_gdata;
+struct crt_grp_gdata;
+
+/* CaRT global data */
+struct crt_gdata {
+	crt_phy_addr_t		cg_addr;
+
+	bool			cg_server;
+	/*
+	 * share NA addr flag, true means all contexts share one NA class, fasle
+	 * means each context has its own NA class.  Each NA class has an
+	 * independent listening address.
+	 */
+	bool			cg_share_na;
+	int			cg_na_plugin; /* NA plugin type */
+
+	/* global timeout value (second) for all RPCs */
+	uint32_t		cg_timeout;
+	/* credits limitation for #inflight RPCs per target EP CTX */
+	uint32_t		cg_credit_ep_ctx;
+
+	/* CaRT contexts list */
+	d_list_t		cg_ctx_list;
+	/* actual number of items in CaRT contexts list */
+	int			cg_ctx_num;
+	/* maximum number of contexts user wants to create */
+	uint32_t		cg_ctx_max_num;
+	/* the global opcode map */
+	struct crt_opc_map	*cg_opc_map;
+	/* HG level global data */
+	struct crt_hg_gdata	*cg_hg;
+
+	struct crt_grp_gdata	*cg_grp;
+
+	/* refcount to protect crt_init/crt_finalize */
+	volatile unsigned int	cg_refcount;
+
+	/* flags to keep track of states */
+	volatile unsigned int	cg_inited		: 1,
```

Comment:
(style)  Use of volatile is usually wrong: see Documentation/volatile-considered-harmful.txt